src/lab16.py

```diff
@@ -288,15 +315,29 @@ def __init__(self, node, parent, previous, frame):
         self.previous = previous
         self.frame = frame
 
-        self.children = ProtectedField(node, "children", self.parent)
-        self.zoom = ProtectedField(node, "zoom", self.parent)
-        self.width = ProtectedField(node, "width", self.parent)
-        self.height = ProtectedField(node, "height", self.parent)
-        self.x = ProtectedField(node, "x", self.parent)
-        self.y = ProtectedField(node, "y", self.parent)
+        self.children = ProtectedField(self, "children", self.parent)
+        self.zoom = ProtectedField(self, "zoom", self.parent)
+        self.width = ProtectedField(self, "width", self.parent)
+        self.height = ProtectedField(self, "height", self.parent)
+        self.x = ProtectedField(self, "x", self.parent)
+        self.y = ProtectedField(self, "y", self.parent)
 
         self.has_dirty_descendants = True
 
+        # self.children has dependencies on the zoom and font style of various
+        # inline children. But we don't know which until self.recurse is
+        # called, and building up the dependencies would amount to a duplicate
+        # implementation of self.recurse.
+        self.zoom.set_dependencies([self.parent.zoom])
+        self.width.set_dependencies([self.parent.width])
+        self.x.set_dependencies([self.parent.x])
+
+        previous_dependencies = [self.parent.y]
+        if self.previous:
+            previous_dependencies.append(self.previous.y)
+            previous_dependencies.append(self.previous.height)
```

Comment:
How about doing something like:
```
self.y.set_dependencies([
  self.parent.y,
  self.previous and self.previous.y,
  self.previous and self.previous.height,
])
```

This is a bit of type punning (we will need `set_dependencies` to skip `False`) but shorter and, I think, clearer. It also avoids naming the temporary.

Reply:
Also, if there is a `previous`, you actually don't directly depend on the parent's `y` position. I don't think this is a _problem_, per se, since if the parent `y` position changes the previous element's `y` position probably will too, but you could imagine adding CSS features where this over-invalidates. Either way, I think we'll need to mention this in the text.

Reply:
Good idea, done.